src/commands/cmd_list.cc

```diff
@@ -167,11 +167,11 @@ class CommandLMPop : public Commander {
     }
     num_keys_ = *v;
 
-    if ((args.size() != 3 + num_keys_) && (args.size() != 5 + num_keys_)) {
+    if ((args.size() != num_keys_ + 3) && (args.size() != num_keys_ + 5)) {
       return {Status::RedisParseErr, errWrongNumOfArguments};
     }
 
-    std::string left_or_right = util::ToLower(args[1 + num_keys_]);
+    std::string left_or_right = util::ToLower(args[num_keys_ + 2]);
```

Comment:
@HolyLow Can use the CommandParser to parse arguments.

Reply:
@git-hulk I have some questions on the usage of CommandParser.
The common usage of CommandParser is like:
`
while (cmdParser.isGood()) {
  if (eatCase0()) { doSomething(); }
  else if (eatCase1()) { doSomething(); }
  else { return ParseError(kInvalidSyntax); }
}
`
But this schema assumes no strict order constraint and no strict repeat constraint. For example, the standard LMPOP command is like: 

*LMPOP 2 key1 key2 LEFT COUNT 2*; 

but with the parsing schema above, this command is also considered as valid: 

*LMPOP 2 key1 key2 LEFT RIGHT COUNT 2 COUNT 4 RIGHT*. 

The repeat and the order of keywords are not constrained by the isGood() loop.  Even though we could constrain the repeat by checking args.size(), the order of keywords is not constrained.

So should I constrain the order of the keywords, or leave it loose? If the orders of the keywords SHOULD be constrained, the existing parsers using CommandParser might have parsing bugs, as the CommandParser has no order semantic by design.

Reply:
We can align the behavior with Redis, I'm not sure if Redis allows the duplicate keyword like you mentioned.

> The repeat and the order of keywords are not constrained by the isGood() loop.

We can work around this by checking if the variable has already been set. For example:

```
direction = DirectionNone
while (cmdParser.isGood()) {
  if (eatCase0()) { doSomething(); }
  else if (eatCase1("LEFT")) {
     if (direction != DirectionNone) {
        // throw error
     }
     direction = DirectionLeft
  } else { return ParseError(kInvalidSyntax); }
}
```

Reply:
Sure, we could find out some way to handle the order/repeat issue, but I just think that these workarounds lose the elegancy brought by CommandParser. Maybe the better way is to refactor CommandParser to support order/repeat checking semantic (which is painful...). Well this might be another issue.

I would refactor the parsing with the CommandParser and check if redis cares about keyword repeat/order.

Reply:
IMHO here we do not need a `while` loop.

```c++
CommandParser parser(args, 1);

auto num_keys = GET_OR_RET(parser.TakeInt<uint32_t>());
for (uint32_t i = 0; i < num_keys; ++i) keys_.emplace(GET_OR_RET(parser.TakeStr()));

auto left_or_right = GET_OR_RET(parser.TakeStr());
// do something about left_or_right 

if (parser.Good()) {
  if (parser.EatEqICase("count")) count_ = GET_OR_RET(parser.TakeInt<uint32_t>());
  else return parser.InvalidSyntax();
}

if (parser.Good()) return parser.InvalidSyntax();
```

To your questions:

> as the CommandParser has no order semantic by design.

Nope. Whether there is an order is entirely up to you, not the CommandParser.

Your code contains so many magic index numbers, which makes it very hard to maintain and extend.

Reply:
Thank you! Redis should not care about the order in most commands if my memory works. cc @PragmaTwice

Reply:
@PragmaTwice you are right about the order issue, it's the user's responsibility to charge for the order rather than CommandParser. Besides, the GET_OR_RET macro makes the err handling much easier.

@git-hulk I've refactored the code and tested the redis behavior. Seems for most commands (e.g. LPOS) the ordering of different keywords doesn't matter, but for LMPOP, the ordering must be fixed, otherwise redis reports "syntax error". So our discussion here is not for nothing.